moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,572 @@
+        # what xdmod is expecting:
+        "instance_type_id": "9",
+        "launched_at": launched_ts,
+        # "deleted_at": terminated_ts,
+        "memory_mb": str(flavor_dict[server.flavor["id"]]["ram"]),
+        # Probably unknowable - I suspect this is what ceilometer adds.
+        # "message_id": "17d6645e-90f9-481e-9751-f8ec3b9397a2",
+        "project_id": server.tenant_id,
+        "project_name": project_dict[server.tenant_id]["name"],
+        "raw": {},
+        # request_id can be NULL
+        # "request_id": "req-00b3079e-8cb1-4e63-aa6f-f96fcbd4771c",
+        # For compute evens, resource_id is the instance id
+        "resource_id": server.id,
+        "root_gb": str(flavor_dict[server.flavor["id"]]["disk"]),
+        "service": "compute",
+        "state": server.status,
+        "tenant_id": server.tenant_id,
+        "user_id": server.user_id,
+        "user_name": user_name,
+        "vcpus": str(flavor_dict[server.flavor["id"]]["vcpus"]),
+    }
+    if terminated_ts is not None:
+        server_state["deleted_at"] = terminated_ts
+    return server_state
+
+
+def build_event(server_state, event):
+    server_event = copy.deepcopy(server_state)
+    if event["event_type"] == "compute.instance.exists":
+        server_event["audit_period_beginning"] = copy.copy(event["audit_period_start"])
+        server_event["audit_period_ending"] = copy.copy(event["audit_period_end"])
+    server_event["event_type"] = copy.copy(event["event_type"])
+    server_event["generated"] = copy.copy(event["event_time"])
+    return server_event
+
+
+def get_session_for_resource(resource):
+    auth_url = resource.get_attribute(attributes.RESOURCE_AUTH_URL)
+    # Note: Authentication for a specific OpenStack cloud is stored in env
+    # variables of the form OPENSTACK_{RESOURCE_NAME}_APPLICATION_CREDENTIAL_ID
+    # and OPENSTACK_{RESOURCE_NAME}_APPLICATION_CREDENTIAL_SECRET
+    # where resource name is has spaces replaced with underscored and is
+    # uppercase.
+    # This allows for the possibility of managing multiple OpenStack clouds
+    # via multiple resources.
+    var_name = resource.name.replace(" ", "_").replace("-", "_").upper()
+    auth = v3.ApplicationCredential(
+        auth_url=auth_url,
+        application_credential_id=os.environ.get(f"OPENSTACK_{var_name}_APPLICATION_CREDENTIAL_ID"),
+        application_credential_secret=os.environ.get(f"OPENSTACK_{var_name}_APPLICATION_CREDENTIAL_SECRET"),
+    )
+    return session.Session(auth, verify=os.environ.get("FUNCTIONAL_TESTS", "") != "True")
+
+
+def main():
+    """The main function"""
+    config = {}
+
+    do_parse_args(config)
+    do_read_config(config)
+
+    openstack_conn = openstack.connect(cloud=config["cloud"])
+    openstack_nova = nova_client.Client(2, session=openstack_conn.session)
+    openstack_cinder = cinder_client.Client(3, session=openstack_conn.session)
+    openstack_keystone = keystone_client.Client(3, session=openstack_conn.session)
```

Comment:
This variable isn't used anywhere.

Reply:
Yes, I used it at one point, but then didn't really need it.  Just didn't get around to eliminating it here.  Will do so.